dsp/modules/premai.py

```diff
@@ -122,8 +120,31 @@ def _get_all_kwargs(self, **kwargs) -> Dict[str, Any]:
     def basic_request(self, prompt, **kwargs) -> str:
         """Handles retrieval of completions from Prem AI whilst handling API errors."""
         all_kwargs = self._get_all_kwargs(**kwargs)
-        messages = []        
-        messages.append({"role": "user", "content": prompt})
+
+        if "template_id" not in all_kwargs:
+            messages = [{"role": "user", "content": prompt}]
+        else:
+            template_id = all_kwargs["template_id"]
+            if (template_id is None) or (template_id == ""):
+                raise ValueError("Templates can not be None or ''")
+            if "params" not in all_kwargs:
+                raise KeyError(
+                    "Keyword argument: params must be present if template_id is present",
+                )
+
+            params = all_kwargs["params"]
+            if not isinstance(params, dict):
+                raise TypeError("params must be a dictionary")
+
+            messages = [
+                {
+                    "role": "user",
+                    "template_id": template_id,
+                    "params": params,
+                },
+            ]
+            all_kwargs.pop("template_id")
```

Comment:
any reason to remove the params from the kwargs which are propagated in history? might be relevant to see how the DSPy prompt was changed in addition to the Prem template but I'll defer to you here.